From bd5485fe266a649d89bfec24c7bcdd6f4553f465 Mon Sep 17 00:00:00 2001
From: Hidehiko Abe <hidehiko@chromium.org>
Date: Sat, 22 Sep 2018 02:27:15 +0900
Subject: [PATCH] Add Optional compatible methods.

Added Optional compatible methods to String for "string?".

BUG=b:37434548
TEST=Build.
---
 mojo/public/cpp/bindings/string.h | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/mojo/public/cpp/bindings/string.h b/mojo/public/cpp/bindings/string.h
index 8bb07be..373b187 100644
--- a/mojo/public/cpp/bindings/string.h
+++ b/mojo/public/cpp/bindings/string.h
@@ -96,6 +96,17 @@ class String {
   const std::string& get() const { return value_; }
   operator const std::string&() const { return value_; }
 
+  template <typename U>
+  std::string value_or(U&& default_value) const& {
+    return is_null_ ? std::string(std::forward<U>(default_value)) : value_;
+  }
+
+  template <typename U>
+  std::string value_or(U&& default_value) && {
+    return is_null_ ?
+        std::string(std::forward<U>(default_value)) : std::move(value_);
+  }
+
   // Returns a const reference to the |std::string| managed by this class. If
   // the string is null, this will be an empty std::string.
   const std::string& storage() const { return value_; }
-- 
2.19.0.444.g18242da7ef-goog

